### PR TITLE
WIP: realtek: Enable 6.1 testing

### PR DESCRIPTION
**Pull request by oliv3r**
With the landing of 6.1, we can now start migrating the realtek target to 6.1 as well.

This MR includes https://github.com/openwrt/openwrt/pull/12725 as we are dependent on it, once that one is merged, a rebase will remove these patches from this queue automatically :)

### CODE REVIEW

**Comment by oliv3r** [comment]
WIP notes:
* not sure what the new DSA changes mean. What do we need to add/change.
* Getting an error right now `irq: no irq domain found for interrupt-controller@3000` no idea what causes this (mangled patch?) but it breaks the timer after, so won't boot atm.
* everything else ...

**Comment by oliv3r** [comment]
So I found out, that 6.1 returns a null pointer for `of_node_to_fwnode(node)` that we used to rely on. Any idea what could cause that/is the reason for that? Do we manually have to fill it? something to learn/investigate ...

**Comment by oliv3r** [comment]
> Well, all that `of_node_to_fwnode()` does is return `&np->fwnode` if it exists, and if not just returns `NULL`. So you gotta investigate what is supposed to attach the fwnode in the first place.

Yeah, turns out, afaik that fwnode is optional and not relevant. That check is also not needed in the select function.

But for some reason, I'm passing a nullptr for the output thing, just curious why it worked in the past of course, so debugging that.

File: target/linux/realtek/rtl838x/config-6.1 [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+CONFIG_CRYPTO_RNG2=y
+CONFIG_DEBUG_SECTION_MISMATCH=y
+CONFIG_DMA_NONCOHERENT=y
+CONFIG_DTC=y
+CONFIG_EARLY_PRINTK=y
+CONFIG_EARLY_PRINTK_8250=y
+CONFIG_EXTRA_FIRMWARE="rtl838x_phy/rtl838x_8214fc.fw rtl838x_phy/rtl838x_8218b.fw rtl838x_phy/rtl838x_8380.fw"
+CONFIG_EXTRA_FIRMWARE_DIR="firmware"
+CONFIG_FIXED_PHY=y
+CONFIG_FORCE_MAX_ZONEORDER=13
```
Comment (stklcode) [inline target/linux/realtek/rtl838x/config-6.1:57]: `FORCE_MAX_ZONEORDER` should be renamed to `ARCH_FORCE_MAX_ORDER` for 6.1

Applies to all 4 _config-6.1_ files

see https://github.com/torvalds/linux/commit/0192445cb2f7ed1cd7a95a0fc8c7645480baba25

**Comment by Neustradamus** [comment]
For information: Kernel 6.6 is LTS:
6.6 | Greg Kroah-Hartman & Sasha Levin | 2023-10-29 | Dec, 2026
- https://www.kernel.org/category/releases.html

**Comment by oliv3r** [comment]
Probably best to drop this MR and swich to 6.6

**Comment by svanheule** [comment]
Since 23.05 is 5.15-based, so the next release will be 6.1. We still need this I'm afraid.

Trying to compile the changes for rtl838x, not all the patches are applying as they are and 703 is even failing:
```
Applying /home/sander/projects/org.openwrt/openwrt/target/linux/realtek/patches-6.1/300-mips-add-rtl838x-platform.patch using plaintext: 
patching file arch/mips/Kbuild.platforms
patching file arch/mips/Kconfig
Hunk #1 succeeded at 990 (offset 3 lines).

[...]

Applying /home/sander/projects/org.openwrt/openwrt/target/linux/realtek/patches-6.1/302-clocksource-add-otto-driver.patch using plaintext: 
patching file drivers/clocksource/Kconfig
patching file drivers/clocksource/Makefile
patching file include/linux/cpuhotplug.h
Hunk #1 succeeded at 179 (offset -1 lines).

[...]

Applying /home/sander/projects/org.openwrt/openwrt/target/linux/realtek/patches-6.1/318-add-rtl83xx-clk-support.patch using plaintext: 
patching file drivers/clk/Kconfig
Hunk #1 succeeded at 454 (offset 1 line).
patching file drivers/clk/Makefile

Applying /home/sander/projects/org.openwrt/openwrt/target/linux/realtek/patches-6.1/700-net-dsa-add-support-for-rtl838x-switch.patch using plaintext: 
patching file drivers/net/dsa/Kconfig
Hunk #1 succeeded at 87 (offset 24 lines).
patching file drivers/net/dsa/Makefile
Hunk #1 succeeded at 24 (offset 2 lines).

[...]

Applying /home/sander/projects/org.openwrt/openwrt/target/linux/realtek/patches-6.1/703-include-linux-add-phy-ops-for-rtl838x.patch using plaintext: 
patching file include/linux/phy.h
Hunk #1 FAILED at 978.
1 out of 1 hunk FAILED -- saving rejects to file include/linux/phy.h.rej
Patch failed!  Please fix /home/sander/projects/org.openwrt/openwrt/target/linux/realtek/patches-6.1/703-include-linux-add-phy-ops-for-rtl838x.patch!
```

@musashino205 I can't find you wiki right now, but I don't suppose you've looked into bumping the kernel for realtek? Just asking, not making any requests here :-)

**Comment by musashino205** [comment]
@svanheule I found this PR just as I was about to start working on bumping, so I put mine on hold.

**Comment by oliv3r** [comment]
@musashino205 probably best if we skip 6.1 and go to 6.6 once openwrt enables support (I haven't checked). But best to work together :) I think this MR was in quite a good shape, except for the non-working part :) (I meant the patches :p)

What do you think?

**Comment by Neustradamus** [comment]
@oliv3r: I have done a ticket to inform OpenWrt team to upgrade master to Kernel 6.6.x LTS but my ticket has been closed.
To be good, I think that OpenWrt must change the strategy and release the next version with the latest Kernel LTS.

In more, about version, now LibreOffice has changed strategy and currently in beta 2, it has already the good year.month of the final version: https://wiki.documentfoundation.org/ReleaseNotes/24.2

OpenWrt has always a version which is not linked with the month of the first release :/

**Comment by musashino205** [comment]
IMO, we shoud to keep 6.1.
Bumping kernel version requires a huge amount of work and places a heavy burden not only on realtek target, but also on other targets and members of OpenWrt team.
There are many targets which bumping is currently being done or has already been done, so we shouldn't put any more burden on them.

**Comment by oliv3r** [comment]
@Neustradamus why was it closed? I thought the OpenWRT releases where intended to track mainline where possible, and thus use the latest LTS where possible? This confuses me a bit tbh.

@musashino205 well there's a few things here. We will have to go to 6.6 eventually _anyway_. As you said, going from 5.15 to 6.1 is a big hurdle as it is. (Unless fixing this MR is easy :p). So why put in the effort of migrating and fixing 6.1, and then doing the work again for 6.6? Might just skip 6.1 and effort, and jump to 6.6 and do the work once.

Of course, we can't postpone these things forever, or we end up nowhere. But with 6.6 out since november, we're chasing a sailed ship with 6.1?

On a different note, I'm not sure where I stranded. Could be because I lack hardware, and I need additional patches to make my hardware work. Also, there where some changes with the PHY stuff, which required quite a bit of work ...

**Comment by svanheule** [comment]
6.1 needs to be done first for 24.xx. Work on 6.6 can be started after 24.xx has been forked.

**Comment by svanheule** [comment]
I've taken this PR, stripped out the unrelated changes (the broken interrupts were due to a bad refactor), cleaned up a bit, and pushed the changes to a staging branch:
https://git.openwrt.org/?p=openwrt/staging/svanheule.git;a=shortlog;h=refs/heads/realtek/testing-6.1

Getting reliable a network traffic on rtl930x isn't there yet, as described in the commit message adding the testing kernel. If anyone can quickly figure out what's wrong, it can still be fixed before pushing this to main in a few days. Otherwise it'll be a testing kernel for a reason ;-)

**Comment by oliv3r** [comment]
> I've taken this PR, stripped out the unrelated changes (the broken interrupts were due to a bad refactor), cleaned up a bit, and pushed the changes to a staging branch: https://git.openwrt.org/?p=openwrt/staging/svanheule.git;a=shortlog;h=refs/heads/realtek/testing-6.1
> 
> Getting reliable a network traffic on rtl930x isn't there yet, as described in the commit message adding the testing kernel. If anyone can quickly figure out what's wrong, it can still be fixed before pushing this to main in a few days. Otherwise it'll be a testing kernel for a reason ;-)

Great news!

Not sure which patches where 'unrelated' as I think I did stripping down as much as I could, of course I am testing on an rtl930x target, which may have required some patches.

I'm surprised you introduced `https://git.openwrt.org/?p=openwrt/staging/svanheule.git;a=commit;h=a7a4b7ca7834125daac4fb145b7e2aa4a597ca89` as this was introduced with 5.15 as well, but strong consensus was, to NOT do this, as the risk of breaking things/backwards compatibility etc was to big. What made you decide that it was a good idea after al?

**Comment by oliv3r** [comment]
Looks like my suggestion of 6.6 was being shared by others. http://lists.openwrt.org/pipermail/openwrt-devel/2024-February/042204.html as posted elsewhere also.

**Comment by robimarko** [comment]
Any news on this or 6.6?

Because, realtek is basically the last struggler and I want to remove 5.15 sooner rather than later

**Comment by oliv3r** [comment]
Sorry for being a bit awol; i wasn't aware you guys had progressed so much!

**Comment by piaddic7ed** [comment]
> Sorry for being a bit awol; i wasn't aware you guys had progressed so much! It's been on my (short) todo list!

Drop me a message if you need someone to test any of your dev/next/etc. branches.

**Comment by robimarko** [comment]
Can you please work on this as realtek is the only target using 5.15 and I want to get rid of 5.15 ASAP.

**Comment by SimonFischer04** [comment]
> Can you please work on this as realtek is the only target using 5.15 and I want to get rid of 5.15 ASAP.

How about you don't demand something from volunteers?
Contribute yourself. Ore if you are not able to yourself, pay someone to do it.

**Comment by robimarko** [comment]
> > Can you please work on this as realtek is the only target using 5.15 and I want to get rid of 5.15 ASAP.
> 
> How about you don't demand something from volunteers? Contribute yourself. Ore if you are not able to yourself, pay someone to do it.

In the future please look at the project commit history before writing things like these.
We currently have to maintain 5.15 support in OpenWrt just for `realtek` and I can tell you that nobody is interested in that as all targets have had more than enough time to move off 5.15.

**Comment by SimonFischer04** [comment]
> > > Can you please work on this as realtek is the only target using 5.15 and I want to get rid of 5.15 ASAP.
> > 
> > 
> > How about you don't demand something from volunteers? Contribute yourself. Ore if you are not able to yourself, pay someone to do it.
> 
> In the future please look at the project commit history before writing things like these. We currently have to maintain 5.15 support in OpenWrt just for `realtek` and I can tell you that nobody is interested in that as all targets have had more than enough time to move off 5.15.

Sorry. 
Maybe I got a little pissed off yesterday after reading multiple threads with the classic "+1" / "any progress" / ... comments and people that actually just complain, so I did not properly check your history.

**Comment by howels** [comment]
> Sorry for being a bit awol; i wasn't aware you guys had progressed so much! It's been on my (short) todo list!

Let me know if you need testers, I have been looking to add the HPE JG927A realtek target (https://github.com/openwrt/openwrt/pull/15763#) but looks like this is taking priority ahead of other new realtek targets.

**Comment by oliv3r** [comment]
Sadly, I won't have any time to work on this for at least a few more weeks.

While sad, I think it's wise to drop the realtek target for now. It's hanging on a thread, and quite a bit messy. A lot of work is in dire need to be done; and has stalled. Not helped by our initial hackers leaving.

We can always re-add it when we have a proper MR to put things in better shape/perspective?

WDY @robimarko ?

**Comment by howels** [comment]
> Sadly, I won't have any time to work on this for at least a few more weeks.
> 
> While sad, I think it's wise to drop the realtek target for now. It's hanging on a thread, and quite a bit messy. A lot of work is in dire need to be done; and has stalled. Not helped by our initial hackers leaving.
> 
> We can always re-add it when we have a proper MR to put things in better shape/perspective?
> 
> WDY @robimarko ?

That would be sad, as almost all of the pure switch devices with OpenWRT support are realtek targets.

**Comment by oliv3r** [comment]
> > Sadly, I won't have any time to work on this for at least a few more weeks.
> > While sad, I think it's wise to drop the realtek target for now. It's hanging on a thread, and quite a bit messy. A lot of work is in dire need to be done; and has stalled. Not helped by our initial hackers leaving.
> > We can always re-add it when we have a proper MR to put things in better shape/perspective?
> > WDY @robimarko ?
> 
> That would be sad, as almost all of the pure switch devices with OpenWRT support are realtek targets.

I'd almost say 'only'; but the realtek target is in a really bad shape if you ask me, and if you have one of the 'prime supported devices' e.g. the first gen of the rtl8xxx series you are a bit better. But we where dicking around too much in getting things perfect, that we kind of steam, and for the time being, things are lost.

I still believe in my patches; but i'll need to dig into shit again; and probably fix some things here or there. And it was quite lonely and desperate for a while; so kind of holding back before burning again :)

**Comment by howels** [comment]
Yes, I have 5 realtek switches on my desk over here! All are RTL83xx and there's only a single 93xx target in master currently. Perhaps 93xx could wait while 83xx is carried forward? I noticed that RTL9302C is the focus of a PR upstream in 6.11 so we might find 93xx in better shape once that reaches a wider audience.

**Comment by oliv3r** [comment]
> Yes, @oliv3r I have 5 realtek switches on my desk over here! All are RTL83xx and there's only a single 9xxx target in master currently. Perhaps 9xxx could wait while 83xx is carried forward? I noticed that [RTL9302C is the focus of a PR upstream in 6.11](https://lore.kernel.org/lkml/ZpuAyzljTBjd6a7g@alpha.franken.de/) so we might find 9xxx in better shape and be able to port patches more easily once that reaches a wider audience.

Can you link that upstream PR?

But I wouldn't hold my breath. Upstream is also in a poor shape, and it does only half of what we need. Also, upstream just does some very very basic stuff.

the biggest 'problem' is actually a lot of the networking stuff itself. It doesn't meet the OpenWRT standard of quality, which is fair, but improving it becomes really hard, if new stuff must adhere, because when refactoring, you have to make compromises.

I tried once to pull the stuff apart and see if we could make it loadable modules. Sadly, this simply didn't work for the networking bits, as it was so intertwined.

The rtl target should never have been in there with its current quality. So either we drop it. Allow lesser coding standards in the hopes it improves, or revisit the 'out of tree' solution again. Maintaining a fork however, while an option, is a bit too much effort sadly ...

**Comment by howels** [comment]
> > Yes, @oliv3r I have 5 realtek switches on my desk over here! All are RTL83xx and there's only a single 9xxx target in master currently. Perhaps 9xxx could wait while 83xx is carried forward? I noticed that [RTL9302C is the focus of a PR upstream in 6.11](https://lore.kernel.org/lkml/ZpuAyzljTBjd6a7g@alpha.franken.de/) so we might find 9xxx in better shape and be able to port patches more easily once that reaches a wider audience.
> 
> Can you link that upstream PR?
> 
> But I wouldn't hold my breath. Upstream is also in a poor shape, and it does only half of what we need. Also, upstream just does some very very basic stuff.
> 
> the biggest 'problem' is actually a lot of the networking stuff itself. It doesn't meet the OpenWRT standard of quality, which is fair, but improving it becomes really hard, if new stuff must adhere, because when refactoring, you have to make compromises.
> 
> I tried once to pull the stuff apart and see if we could make it loadable modules. Sadly, this simply didn't work for the networking bits, as it was so intertwined.
> 
> The rtl target should never have been in there with its current quality. So either we drop it. Allow lesser coding standards in the hopes it improves, or revisit the 'out of tree' solution again. Maintaining a fork however, while an option, is a bit too much effort sadly ...

Sounds rough, and clearly unsustainable without a larger number of maintainers.
The upstream patch is https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=d2be38b9a5514dbc7dc0c96a2a7f619fcddce00d

I tried to rework the patches for 6.6 last night. I might be able to get something that compiles but when it comes to handling IRQ hardware I'm not sure what I'm doing so doubt the results will work properly, much less improve quality unfortunately.

**Comment by robimarko** [comment]
@oliv3r I agree that `realtek` is in a rather poor shape, especially all of the more modern switch IC-s in the target.
I think it was pushed into OpenWrt way too early and now we are stuck with a mountain of unmaintainable code.

As far as dropping it goes, that would be rather sad as its quite popular and the only pure switch target we have.
But, the truth is that if it can't be updated to 6.6 (That is the next release kernel), we won't have a choice but to drop it.

That being said, this should be posted on the development list as a question so a wider audience could comment and maybe somebody can offer some man-hours to at least get another OpenWrt release out of it.

So, I urge anybody who needs this target to offer some help as we honestly cant keep having 5.15 just for the sake of this target.

**Comment by svanheule** [comment]
Closing as superseded by https://github.com/openwrt/openwrt/pull/16204

**Pull request closed without merging**